### DISCUSSION TITLE
multiple objects squeeze each other

### DISCUSSION THREAD

**Discussion by dannisliang** [bepu/bepuphysics1#27] [Q&A]
[http://www.apolloads.cn/20220427095357.png](url)
In the V1 version, when the multiple objects squeeze each other, the object in the middle will rise upward and the y-axis will also increase. How can these be avoided?

**Comment by RossNordby** [comment]
If the magnitude of penetration is great enough, it's possible that going up is the shortest path to separation.

The picture looks like the penetration magnitudes are too small for this to be happening, though. If I had to guess, the solver can't converge to a solution for the collision constraints, so an intermediate solution ends up with tangential velocity and squeezes out sideways. Try setting `Space.TimeStepSettings.TimeStepDuration` to a value 10x smaller, and then calling `Space.Update` more frequently to compensate. Then, set `Space.Solver.IterationLimit` to a high value like 50, and `SolverSettings.DefaultMinimumIterationCount = int.MaxValue`. If that seems to resolve the issue, it's because the solver is having trouble converging. 

If that's the cause, then using a shorter time step duration or using more solver iterations is a good choice. It is unlikely you will need to go to the extremes of the above diagnostic- it could be that just increasing the iteration limit to 12 or something alone fixes the issue. Just pick the cheapest option that works. Increasing `CollisionResponseSettings.Softness` may help too; it makes the solver's job easier.

Also, check friction. If friction is too low, then any slight tangential motion will be preserved and things will tend to slide.

**Reply by dannisliang** [reply]
Thank you for the quick response!

[http://www.apolloads.cn/20220428094421.mp4](url)

[http://www.apolloads.cn/20220428095035.png](url)

I found that it is caused by entityA.ApplyAngularImpulse(ref angular)，entityB.ApplyAngularImpulse(ref angular) in the ExclusiveUpdate() function in ContactPenetrationConstraint.cs, is there a way to prevent this？Or not execute ApplyAngularImpulse?

**Reply by RossNordby** [reply]
That's unlikely to be the right level to deal with the issue- that's a core piece of solver functionality. Disabling it will break a lot of stuff.

If that *appears* to help, it's likely that the problem is related to the solver failing to converge, or possibly a matter of how the objects' motion is being controlled.

Note that if positions/orientations are set directly, their velocities are *not* updated to match the resulting motion, which means collision constraints will be unable to stop the motion. Setting position/orientation directly is equivalent to teleporting the object. Controlling the motion of objects by modifying their velocities (or using velocity constraints, if necessary) is advised.

If you already are controlling bodies with velocity or constraints, then it's still probably the solver having difficulty converging and the earlier recommendations hold.

If you don't want objects to rotate at all, set `entity.LocalInertiaTensorInverse = default`. That gives the body effectively infinite rotational inertia, so no forces can rotate it.

**Reply by dannisliang** [reply]
Thank you for the quick response!

I found that setting the InertiaHelper.InertiaTensorScale=0 can solve this problem. But the setting of InertiaTensorScale value is a global setting. Is there any way to set it to 0 for an individual entity?

**Reply by dannisliang** [reply]
It seems that setting LinearVelocity while setting LocalInertiaTensorInverse  can also be solved。
public FixVector3 RoleMove
        {
            set
            {
                entity.LinearVelocity= value;
                entity.LocalInertiaTensorInverse = new Matrix3x3();
            }
        }

**Reply by RossNordby** [reply]
>But the setting of InertiaTensorScale value is a global setting. Is there any way to set it to 0 for an individual entity?

That's equivalent to `entity.LocalInertiaTensorInverse = new Matrix3x3()`.

Notably, you only need to set the local inertia tensor inverse once per entity; it does not change by itself.

**Reply by dannisliang** [reply]
Many thanks for your instruction.
